runtime/lua/vim/diagnostic.lua

```diff
@@ -706,7 +706,7 @@ local function get_diagnostics(bufnr, opts, clamp)
   ---@param b integer
   ---@param d vim.Diagnostic
   local function add(b, d)
-    if not opts.lnum or d.lnum == opts.lnum then
+    if not opts.lnum or d.lnum == opts.lnum or d.end_lnum == opts.lnum then
```

Comment:
Rather than comparing directly to `lnum` or `end_lnum`, `opts.lnum` should get any diagnostic that spans that line number. So it should be (I think):

```lua
(opts.lnum >= d.lnum) and (opts.lnum <= (d.end_lnum or d.lnum))
``` 

So if a diagnostic has _only_ a `lnum` (no `end_lnum`) this works as before. If `lnum` and `end_lnum` are equal, this works as before. But if `end_lnum > lnum`, then any `opts.lnum` that lies within the range would be returned.

The docstring for `GetOpts` should be updated too.